clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -1076,29 +1086,36 @@ public void run() {
                 }
 
                 // Choose nodes for our pending calls.
-                pollTimeout = Math.min(pollTimeout, chooseNodesForPendingCalls(now, pendingCalls.iterator()));
+                pollTimeout = Math.min(pollTimeout, checkPendingCalls(now));
                 long metadataFetchDelayMs = metadataManager.metadataFetchDelayMs(now);
                 if (metadataFetchDelayMs == 0) {
                     metadataManager.transitionToUpdatePending(now);
                     Call metadataCall = makeMetadataCall(now);
                     // Create a new metadata fetch call and add it to the end of pendingCalls.
                     // Assign a node for just the new call (we handled the other pending nodes above).
-                    pendingCalls.add(metadataCall);
-                    chooseNodesForPendingCalls(now, pendingCalls.listIterator(pendingCalls.size() - 1));
+
+                    if (!checkPendingCall(metadataCall, now))
```

Comment:
Would the logic be slightly simpler by just doing `pendingCalls.add(metadataCall)` here without invoking `checkPendingCall`, and move `pollTimeout = Math.min(pollTimeout, checkPendingCalls(now))` to be after the `if (metadataFetchDelayMs == 0) {...}`?

Reply:
I think the reason it is done in this order is that the node selection itself may trigger the metadata update. What if we just add the metadata call to `pendingCalls` and `continue`?

Reply:
Ah I see. Makes sense. Then I think the current code looks good.